[WGSL] Initial implementation of operator overload resolution

tadeuzagallo opened the pull request:
#### 387c1b1f6bd7c9ae77a9e1ce3c72d85697a14497
<pre>
[WGSL] Initial implementation of operator overload resolution
<a href="https://bugs.webkit.org/show_bug.cgi?id=251863">https://bugs.webkit.org/show_bug.cgi?id=251863</a>
rdar://105133289

Reviewed by Myles C. Maxfield.

First pass at adding support for overloaded operators and functions. For now,
only `+` is implemented with 2 operators, and declaring the operations is very
cumbersome, but the patch is fairly large as it is, so more convenient declarations
will follow in a separate patch.

* Source/WebGPU/WGSL/Overload.cpp: Added.
(WGSL::log):
(WGSL::logLn):
(WGSL::OverloadResolver::OverloadResolver):
(WGSL::OverloadResolver::resolve):
(WGSL::OverloadResolver::materialize const):
(WGSL::OverloadResolver::considerCandidates):
(WGSL::OverloadResolver::considerCandidate):
(WGSL::OverloadResolver::calculateCost):
(WGSL::OverloadResolver::unify):
(WGSL::OverloadResolver::assign):
(WGSL::OverloadResolver::resolve const):
(WGSL::OverloadResolver::conversionCost const):
(WGSL::primitive_pair):
(WGSL::OverloadResolver::conversionCostImpl const):
(WGSL::resolveOverloads):
(WTF::printInternal):
* Source/WebGPU/WGSL/Overload.h: Copied from Source/WebGPU/WGSL/Types.h.
* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::TypeChecker):
(WGSL::TypeChecker::visit):
(WGSL::TypeChecker::chooseOverload):
* Source/WebGPU/WGSL/TypeStore.cpp:
(WGSL::TypeStore::TypeStore):
(WGSL::TypeStore::vectorType):
(WGSL::TypeStore::matrixType):
(WGSL::TypeStore::allocateConstructor):
* Source/WebGPU/WGSL/TypeStore.h:
* Source/WebGPU/WGSL/Types.h:
* Source/WebGPU/WGSL/WGSL.h:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/260254@main">https://commits.webkit.org/260254@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/06571ff0d480be21672fa7c63ffbc05c41d07ff0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107701 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16754 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40596 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116848 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116238 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18175 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8072 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99876 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113455 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13721 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96917 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41401 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/95634 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28558 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83235 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9733 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29906 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10412 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6794 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15886 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49492 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7088 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11973 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WGSL/Overload.cpp:
```diff
@@ -0,0 +1,534 @@
+inline void log(Arguments&&... arguments)
+{
+    if (shouldDumpOverloadDebugInformation)
+        dataLog(logPrefix, std::forward<Arguments>(arguments)...);
```
lol...

litherum commented on Source/WebGPU/WGSL/Overload.cpp:
```diff
@@ -0,0 +1,534 @@
+
+    FixedVector<std::optional<ViableOverload>> considerCandidates();
+    std::optional<ViableOverload> considerCandidate(const OverloadCandidate&);
+    unsigned calculateCost(const AbstractType&, Type*);
```
The way overload resolution works in WGSL doesn't have a cost function.

Firstly, only built-in functions are allowed to overload. User-written functions can't overload.

Secondly, all the built-in functions which are overloaded have mutually-exclusive signatures. There's never any situation where 2 function declarations are valid for a single call.

If you have something like `sin(3.0)`, the literal 3.0 gets concretized according to concretization rules, which are not sensitive to the context that it's being passed to a function. So there's only a single correct overload after the concretization.

tadeuzagallo replied:
That's interesting. I tried to match the spec as closely as possible as described here: https://www.w3.org/TR/WGSL/#overload-resolution-section

> Firstly, only built-in functions are allowed to overload.

This function doesn't assume anything about the overloads being user defined or builtins, but I don't see how assuming they are builtins wouldn't make the resolution any simpler.

> Secondly, all the built-in functions which are overloaded have mutually-exclusive signatures. There's never any situation where 2 function declarations are valid for a single call.

That's true, but that doesn't mean that the same declaration can't yield more than one resolution. i.e. `+ (T, T) -> T` apply to `1 + 1` can resolve `T` to `AbstractInt`, `i32`, `u32`, `AbstractFloat`, `f32` or `f16`, in that order of preference according to the conversion rank table.

> If you have something like sin(3.0), the literal 3.0 gets concretized according to concretization rules, which are not sensitive to the context that it's being passed to a function. So there's only a single correct overload after the concretization.

You know the spec better than me, but that's not how I interpreted it. The way I understood it, overload resolution doesn't concretize types. e.g. `const x = sin(3.0)`, `x` should have type `AbstractFloat`, but if we use `var y = sin(3.0)`, now `y` should have type `f32`, according to the conversion rank function.

Maybe that's some nuance that I'm missing, and it can be implemented in a simpler way than described in the spec.

litherum replied:
> This function doesn't assume anything about the overloads being user defined or builtins, but I don't see how assuming they are builtins wouldn't make the resolution any simpler.

It's a finite set - every member of every set of overload resolutions is known before we look at the first line of the user program.

> That's interesting. I tried to match the spec as closely as possible as described here: https://www.w3.org/TR/WGSL/#overload-resolution-section

https://gpuweb.github.io/gpuweb/wgsl/#conversion-rank is totally new to me, I had no idea that there was a ranking algorithm. My background knowledge here comes from the fact that the CG wanted to forbid user functions from overriding each other, precisely because (IIRC) they didn't want to write a general ranking algorithm - they wanted to punt on that algorithm until after the initial version of the language. I guess my memory is out-of-date, since it seems pretty clear that a general ranking algorithm exists.

Sorry for the noise, I think your following the spec was the right thing to do.

litherum commented on Source/WebGPU/WGSL/Overload.cpp:
```diff
@@ -0,0 +1,534 @@
+    FixedVector<std::optional<ViableOverload>> considerCandidates();
+    std::optional<ViableOverload> considerCandidate(const OverloadCandidate&);
+    unsigned calculateCost(const AbstractType&, Type*);
+    unsigned conversionCost(Type*, Type*) const;
```
WGSL types don't have implicit conversions, except for literals, so I'm not sure this is the right approach.

tadeuzagallo replied:
This is supposed to match 1:1 the "Conversion Rank" table in the spec: https://www.w3.org/TR/WGSL/#conversion-rank

tadeuzagallo replied:
I should rename this to `conversionRank` if we end up landing this.

litherum commented on Source/WebGPU/WGSL/TypeCheck.cpp:
```diff
@@ -112,6 +116,34 @@ TypeChecker::TypeChecker(ShaderModule& shaderModule)
+    // operator + [T<:Number](T, T) -> T
+    OverloadCandidate plus1;
+    {
+        TypeVariable T { 0, TypeVariable::Number };
+        plus1.typeVariables.append(T);
+        plus1.parameters.append(T);
+        plus1.parameters.append(T);
+        plus1.result = T;
```
Is there a way to represent this as data rather than C++? I assume there are going to be dozens of more stanzas like this. Can we generate them at build time? (or run time?)

tadeuzagallo replied:
Yeah, that's the plan. I'm hoping to do something similar to JSC's `BytecodeList.rb`, where we declare all these in a file and output these C++ definitions at compile time.

litherum commented on Source/WebGPU/WGSL/WGSL.h:
```diff
@@ -47,15 +47,15 @@ class ShaderModule;
     SuccessfulCheck() = delete;
     SuccessfulCheck(SuccessfulCheck&&);
-    SuccessfulCheck(Vector<Warning>&&, UniqueRef<ShaderModule>&&);
+    SuccessfulCheck(WTF::Vector<Warning>&&, UniqueRef<ShaderModule>&&);
```
Can we rejigger our includes to make this not have to be changed?

tadeuzagallo replied:
Yeah, that was 100% a mistake, I'll upload a patch to move the types into their own namespace.

djg replied:
Preferably `WGSL::Type::Vector`

tadeuzagallo replied:
Yes, I moved it in #9976. Only I made it `Types`, because I wanted `Type` to remain the base class that lives in the top-level `WGSL` namespace.

litherum reviewed (approved):
Looks pretty straightforward.

tadeuzagallo committed:
[WGSL] Initial implementation of operator overload resolution
https://bugs.webkit.org/show_bug.cgi?id=251863
rdar://105133289

Reviewed by Myles C. Maxfield.

First pass at adding support for overloaded operators and functions. For now,
only `+` is implemented with 2 operators, and declaring the operations is very
cumbersome, but the patch is fairly large as it is, so more convenient declarations
will follow in a separate patch.

* Source/WebGPU/WGSL/Overload.cpp: Added.
(WGSL::log):
(WGSL::logLn):
(WGSL::OverloadResolver::OverloadResolver):
(WGSL::OverloadResolver::resolve):
(WGSL::OverloadResolver::materialize const):
(WGSL::OverloadResolver::considerCandidates):
(WGSL::OverloadResolver::considerCandidate):
(WGSL::OverloadResolver::calculateCost):
(WGSL::OverloadResolver::unify):
(WGSL::OverloadResolver::assign):
(WGSL::OverloadResolver::resolve const):
(WGSL::OverloadResolver::conversionCost const):
(WGSL::primitive_pair):
(WGSL::OverloadResolver::conversionCostImpl const):
(WGSL::resolveOverloads):
(WTF::printInternal):
* Source/WebGPU/WGSL/Overload.h: Copied from Source/WebGPU/WGSL/Types.h.
* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::TypeChecker):
(WGSL::TypeChecker::visit):
(WGSL::TypeChecker::chooseOverload):
* Source/WebGPU/WGSL/TypeStore.cpp:
(WGSL::TypeStore::TypeStore):
(WGSL::TypeStore::vectorType):
(WGSL::TypeStore::matrixType):
(WGSL::TypeStore::allocateConstructor):
* Source/WebGPU/WGSL/TypeStore.h:
* Source/WebGPU/WGSL/Types.h:
* Source/WebGPU/WGSL/WGSL.h:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/260254@main
